README.rst
=============================
dj-libcloud
=============================

.. image:: https://badge.fury.io/py/dj-libcloud.png
    :target: https://badge.fury.io/py/dj-libcloud

Adds easy python 3 and 2.7 support to Django for management of static assets. This is a wrapper around the excellent `Apache Libcloud`_ library.

.. _`Apache Libcloud`: https://libcloud.apache.org/

Documentation
-------------

The full documentation is at https://dj-libcloud.readthedocs.org.


Quickstart
------------------------

Libcloud verifies server SSL certificates before it lets you do anything. It will search your system for the CA certificate, and if it doesn't find it then it will blow up. See https://libcloud.readthedocs.org/en/latest/other/ssl-certificate-validation.html

Installing CA certificate bundle on Mac OS X::

    # Assuming you are using homebrew for Mac OS X dependency management.
    $ brew install curl-ca-bundle

Install dj-libcloud::

    $ pip install dj-libcloud

Then use it in a project, e.g. for your static files::

    # settings.py

    STATIC_URL = 'https://s3.amazonaws.com/my-assets/'
    STATICFILES_STORAGE = 'djlibcloud.storage.LibCloudStorage'

    LIBCLOUD_PROVIDERS = {
        'default': {
            'type': 'libcloud.storage.types.Provider.S3',
            'user': os.environ.get('AWS_ACCESS_KEY'),
            'key': os.environ.get('AWS_SECRET_KEY'),
            'bucket': 'my-assets',
            'secure': True,
        },
    }

Other LibCloud Providers
------------------------

If you want to use other libcloud providers (Rackspace, Openstack, other AWS centers, et al), please visit:

* The `libcloud list of supported providers`_
* The `dj-libcloud cookbook`_.

.. _`libcloud list of supported providers`: https://libcloud.readthedocs.org/en/latest/storage/supported_providers.html
.. _`dj-libcloud cookbook`: http://dj-libcloud.readthedocs.org/en/latest/cookbook.html

Features
--------

* Works for uploading media assets using Python 3.3 and Django 1.6.
* In theory supports all the backends that libcloud supports.

FAQ
-----

Because you just had to ask.

Why not use dj-static or whitenoise?
++++++++++++++++++++++++++++++++++++++++++++++++++++++

Those are great libraries, but are not what you want when handling user uploaded media.

Why not just update django-storages?
++++++++++++++++++++++++++++++++++++++++++++++++++++++

`libcloud` is awesome and has a dedicated team devoted to it. We can have it do most of the heavy lifting. On the other hand, converting `django-storages` to work with Python 3 looked like too much work. Sometimes you just have to start anew, right?

What storage providers does dj-libcloud support?
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

dj-libcloud is a wrapper around libcloud, meaning it supports all the providers of that library. Check out the `full list of supported providers`_!

.. _`full list of supported providers`: https://libcloud.readthedocs.org/en/latest/storage/supported_providers.html



How can I contribute?
++++++++++++++++++++++++++++++++++++

Please read http://dj-libcloud.readthedocs.org/en/latest/contributing.html

What about compressors like django-pipeline?
++++++++++++++++++++++++++++++++++++++++++++++++++++++

Working on it. Currently the `PipelineCachedCloudStorage` class breaks the second time you run it. See https://github.com/pydanny/dj-libcloud/issues/7

CREDIT
------

Many thanks to Jannis Leidel (@jezdez) for giving me the code to get this started. He's a Django core developer, the master of Django static asset managment, and overall a great great guy.


HISTORY.rst
.. :changelog:

History
-------

0.2.0(2014-06-23)
++++++++++++++++++

* Updated url method to return correct values for Rackspace Cloudfiles, Microsoft Azure, Google Storage, non-US Amazon SV (Thanks Jannis!)
* Improved the cookbook immensely (Thanks Jannis!)
* Add link from README to cookbook
* Fix ``url`` method for Google Storage, Rackspace Cloudfiles and
  Microsoft Azure. Also return the correct value for non-US Amazon S3
  buckets.

0.1.2 (2014-04-24)
++++++++++++++++++

* Confirmed to work with Python 2.7
* Remove django-pipeline specific code from storages
* Add cookbook to docs that includes django-pipeline

0.1.1 (2014-04-21)
++++++++++++++++++

* Fixed second-time run problem by just using LibCloudStorage class
* Made django-pipeline optional
* Removed unnecessary files
* Moved TODO to issue tracker

0.1.0 (2014-04-21)
++++++++++++++++++

* First release on PyPI.
* Frustration over lack of easy media asset support for Django and Python 3.

tests/test_models.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_dj-libcloud
------------

Tests for `dj-libcloud` models module.
"""

import os
import shutil
import unittest

from djlibcloud import storage


class TestDjlibcloud(unittest.TestCase):

    def setUp(self):
        pass

    def test_something(self):
        pass

    def tearDown(self):
        pass


tests/test_helpers.py
# !/usr/bin/env python
# -*- coding: utf-8 -*-

import unittest

from djlibcloud.storage import parse_date


class TestHelpers(unittest.TestCase):

    def test_parse_date(self):

        date = parse_date("Sunday, 06-Nov-94 08:49:37 GMT")
        self.assertEqual(date.day, 6)
        self.assertEqual(date.month, 11)
        self.assertEqual(date.year, 1994)

        # second case <= 68
        date = parse_date("Sunday, 06-Nov-05 08:19:32 GMT")
        self.assertEqual(date.year, 2005)


tests/test_storage.py
# !/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_dj-libcloud
------------

Tests for `dj-libcloud` models module.
"""

import os
import shutil
import unittest
from django.core.files.storage import get_storage_class
from django.conf import settings
from djlibcloud.storage import LibCloudStorage


class TestDjlibcloud(unittest.TestCase):

    storage_class = LibCloudStorage

    def setUp(self):
        pass

    def test_get_filesystem_storage(self):
        """
        get_storage_class returns the class for a storage backend name/path.
        """
        self.assertEqual(
            get_storage_class('djlibcloud.storage.LibCloudStorage'),
            LibCloudStorage)
        self.assertEqual(settings.DEFAULT_LIBCLOUD_PROVIDER, 'local')

    @unittest.expectedFailure
    def test_improperly_configured(self):
        storage = self.storage_class()



djlibcloud/storage.py
# -*- coding: utf-8 -*-

from datetime import datetime, timedelta
from io import BytesIO
import os

from django.conf import settings
from django.core.files.storage import Storage
from django.core.files.base import File
from django.core.exceptions import ImproperlyConfigured
from django.utils.six.moves.urllib.parse import urlparse, urlunparse, urljoin


try:
    from email.utils import parsedate_tz
except ImportError:  # pragma: no cover
    from email.Utils import parsedate_tz

try:
    from libcloud.storage.providers import get_driver
    from libcloud.storage.types import ObjectDoesNotExistError, Provider
except ImportError:
    raise ImproperlyConfigured("Could not load libcloud")


def parse_date(value):
    """Parse one of the following date formats into a datetime object:

    .. sourcecode:: text

        Sun, 06 Nov 1994 08:49:37 GMT ; RFC 822, updated by RFC 1123
        Sunday, 06-Nov-94 08:49:37 GMT ; RFC 850, obsoleted by RFC 1036
        Sun Nov 6 08:49:37 1994 ; ANSI C's asctime() format

    If parsing fails the return value is `None`.

    :param value: a string with a supported date format.
    :return: a :class:`datetime.datetime` object.

    Original copied from werkzeug.http
    """
    if value:
        t = parsedate_tz(value.strip())
        if t is not None:
            try:
                year = t[0]
                # unfortunately that function does not tell us if two digit
                # years were part of the string, or if they were prefixed
                # with two zeroes. So what we do is to assume that 69-99
                # refer to 1900, and everything below to 2000
                if year >= 0 and year <= 68:
                    year += 2000
                elif year >= 69 and year <= 99:
                    year += 1900
                seconds = timedelta(seconds=t[-1] or 0)
                return datetime(*((year,) + t[1:7])) - seconds
            except (ValueError, OverflowError):
                return None


class LibCloudStorage(Storage):
    """
    Django storage derived class using apache libcloud to operate
    on supported providers
    """
    def __init__(self, provider_name=None, option=None):
        if provider_name is None:
            provider_name = getattr(settings,
                                    'DEFAULT_LIBCLOUD_PROVIDER',
                                    'default')

        self.provider = settings.LIBCLOUD_PROVIDERS.get(provider_name)
        if not self.provider:
            raise ImproperlyConfigured(
                'LIBCLOUD_PROVIDERS %s not defined or invalid' % provider_name)
        extra_kwargs = {}
        if 'region' in self.provider:
            extra_kwargs['region'] = self.provider['region']
        try:
            provider_type = self.provider['type']
            if isinstance(provider_type, str):
                module_path, tag = provider_type.rsplit('.', 1)
                if module_path != 'libcloud.storage.types.Provider':
                    raise ValueError("Invalid module path")
                provider_type = getattr(Provider, tag)

            Driver = get_driver(provider_type)
            self.driver = Driver(
                self.provider['user'],
                self.provider['key'],
                **extra_kwargs
            )
        except Exception as e:
            raise ImproperlyConfigured(
                "Unable to create libcloud driver type %s: %s" %
                (self.provider.get('type'), e))
        self.bucket = self.provider['bucket']   # Limit to one container
        self.secure = self.provider.get('secure', False)

    def _get_bucket(self):
        """Helper to get bucket object (libcloud container)"""
        return self.driver.get_container(self.bucket)

    def _clean_name(self, name):
        """Clean name (windows directories)"""
        return os.path.normpath(name).replace('\\', '/')

    def _get_object(self, name):
        """Get object by its name. Return None if object not found"""
        clean_name = self._clean_name(name)
        try:
            return self.driver.get_object(self.bucket, clean_name)
        except ObjectDoesNotExistError:
            return None

    def delete(self, name):
        """Delete object on remote"""
        obj = self._get_object(name)
        if obj:
            return self.driver.delete_object(obj)

    def exists(self, name):
        obj = self._get_object(name)
        return True if obj else False

    def listdir(self, path='/'):
        """Lists the contents of the specified path,
        returning a 2-tuple of lists; the first item being
        directories, the second item being files.
        """
        container = self._get_bucket()
        objects = self.driver.list_container_objects(container)
        path = self._clean_name(path)
        if not path.endswith('/'):
            path = "%s/" % path
        files = []
        dirs = []
        # TOFIX: better algorithm to filter correctly
        # (and not depend on google-storage empty folder naming)
        for o in objects:
            if path == '/':
                if o.name.count('/') == 0:
                    files.append(o.name)
                elif o.name.count('/') == 1:
                    dir_name = o.name[:o.name.index('/')]
                    if dir_name not in dirs:
                        dirs.append(dir_name)
            elif o.name.startswith(path):
                if o.name.count('/') <= path.count('/'):
                    # TOFIX : special case for google storage with empty dir
                    if o.name.endswith('_$folder$'):
                        name = o.name[:-9]
                        name = name[len(path):]
                        dirs.append(name)
                    else:
                        name = o.name[len(path):]
                        files.append(name)
        return (dirs, files)

    def size(self, name):
        obj = self._get_object(name)
        if obj:
            return obj.size
        else:
            return -1

    def url(self, name):
        provider_type = self.provider['type'].lower()
        obj = self._get_object(name)
        if not obj:
            return None
        try:
            # currently only Cloudfiles supports it
            url = self.driver.get_object_cdn_url(obj)
        except NotImplementedError as e:
            object_path = '%s/%s' % (self.bucket, obj.name)
            if 's3' in provider_type:
                base_url = 'http://%s' % self.driver.connection.host
                url = urljoin(base_url, object_path)
            elif 'google' in provider_type:
                url = urljoin('http://storage.googleapis.com', object_path)
            elif 'azure' in provider_type:
                base_url = ('http://%s.blob.core.windows.net' %
                            self.provider['user'])
                url = urljoin(base_url, object_path)
            else:
                raise e
        if self.secure:
            if 'cloudfiles' in provider_type:
                parsed_url = urlparse(url)
                if parsed_url.scheme != 'http':
                    return url
                split_netloc = parsed_url.netloc.split('.')
                split_netloc[1] = 'ssl'
                url = urlunparse(
                    'https',
                    '.'.join(split_netloc),
                    parsed_url.path,
                    parsed_url.params, parsed_url.query,
                    parsed_url.fragment
                )
            if ('s3' in provider_type or
                    'google' in provider_type or
                    'azure' in provider_type):
                url = url.replace('http://', 'https://')
        return url

    def modified_time(self, name):
        obj = self._get_object(name)
        last_modified = obj.extra.get('last_modified')
        if last_modified is None:
            return super(LibCloudStorage, self).modified_time(name)
        else:
            return parse_date(last_modified)

    def _open(self, name, mode='rb'):
        remote_file = LibCloudFile(name, self, mode=mode)
        return remote_file

    def _read(self, name, start_range=None, end_range=None):
        obj = self._get_object(name)
        # TOFIX : we should be able to read chunk by chunk
        return next(self.driver.download_object_as_stream(obj, obj.size))

    def _save(self, name, content):
        extra = {'acl': 'public-read'}
        self.driver.upload_object_via_stream(iter(content.file),
                                             self._get_bucket(),
                                             name, extra=extra)
        return name


class LibCloudPrivateStorage(LibCloudStorage):

    def _save(self, name, content):
        self.driver.upload_object_via_stream(iter(content.file),
                                             self._get_bucket(),
                                             name)
        return name


class LibCloudFile(File):
    """
    File inherited class for libcloud storage objects read and write
    """
    def __init__(self, name, storage, mode):
        self._name = name
        self._storage = storage
        self._mode = mode
        self._is_dirty = False
        self.file = BytesIO()
        self.start_range = 0

    @property
    def size(self):
        if not hasattr(self, '_size'):
            self._size = self._storage.size(self._name)
        return self._size

    def read(self, num_bytes=None):
        if num_bytes is None:
            args = []
            self.start_range = 0
        else:
            args = [self.start_range, self.start_range + num_bytes - 1]
        data = self._storage._read(self._name, *args)
        self.file = BytesIO(data)
        return self.file.getvalue()

    def write(self, content):
        if 'w' not in self._mode:
            raise AttributeError("File was opened for read-only access.")
        self.file = BytesIO(content)
        self._is_dirty = True

    def close(self):
        if self._is_dirty:
            self._storage._save(self._name, self.file)
        self.file.close()
